### ISSUE TITLE
Should PerformanceObserver throw when entryTypes is empty?

### ISSUE THREAD

**Issue by igrigorik** [w3c/performance-timeline#87]
PerformanceObserver is specced to ignore unknown entry types, and to throw a TypeError if the supported list of entries is empty. The former behavior is to enable extensibility, and the latter as a warning that you may be creating a no-op observer. However, there is a tension and inconsistency between the two, and we have two high profile cases where this caused web compat issues:

- https://github.com/webcompat/web-bugs/issues/7491 - Spotify wanted to observe "paint" but does not guard for empty list / raised exception.
- https://github.com/webcompat/web-bugs/issues/9716#issuecomment-328694263 - Google Cloud Console made the same mistake with "longtask". 

I suspect that we should [revisit the logic](url)...

> The observe() method instructs the user agent to register the observer and must run these steps:
> 1. Filter unsupported names within the entryTypes sequence, and replace the entryTypes sequence with the new filtered sequence.
> 2. If the options' entryTypes attribute is an empty sequence, throw a JavaScript TypeError.

As mentioned above, the intent behind step (2) is to warn the developer that they may be registering a zombie observer if none of the types are supported. We do have a separate request to enumerate supported task types (https://github.com/w3c/performance-timeline/issues/77), but I don't think that alone is sufficient to address the issue above — it's not obvious that I should manually check the types before requesting them, especially if unknown ones are ignored, and the gotcha case is when all of them are ignored.

As one option, should we consider dropping step 2? Is there a different way to signal that observer won't be receiving any entries and do we care? From an operational standpoint, it's a no-op object.

**Comment by igrigorik** [comment]
@toddreifsteck @spanicker @rniwa @hiikezoe @JosephPecoraro would love to hear your thoughts and feedback on this one.

**Comment by wisniewskit** [comment]
Perhaps it would be best for the call to `.observe` to return an array of the entryTypes the UA was able to subscribe, so that sites can opt into that knowledge or silently ignore it (as they seem to wish to do so right now)?

**Comment by cvazac** [comment]
I think we should drop step 2 - forcing people to try/catch to feature detect doesn't feel right. 

Returning the filtered list per @wisniewskit would work. Question - if the returned list was empty, would developers still need to `.disconnect()`?

One reason to do this would be to polyfill. Here are my attempts:
[interface check](https://gist.github.com/cvazac/22194645f411a68a50c39c88f05faf2d)
[check filtered entries](https://gist.github.com/cvazac/ad97b8fcac33bb02b3367c8180f71c3b)

Both of those examples use a different method of feature detection, which I think we don't want. I think we should have only **one** way to feature detect, lest we confuse developers and risk inter- and intra- browser discrepancies. "check filtered entries" is certainly less verbose, but I prefer "interface check" because I don't think you should have to register something you don't care about, just to feature detect.

**Comment by JosephPecoraro** [comment]
> I think we should drop step 2 - forcing people to try/catch to feature detect doesn't feel right.

I agree.

To state the obvious, a PerformanceObserver that is observing nothing is harmless. It will just do nothing, and that is completely okay. Doing nothing / ignoring the unknown is a common and useful approach for being future proof.

For example: CSS ignores properties it doesn't understand which has been crucial for its evolution. Imagine if `addEventListener` threw an exception if the event name was unsupported.

> the intent behind step (2) is to warn the developer that they may be registering a zombie observer if none of the types are supported.

Exceptions should be for exceptional cases, most commonly unrecoverable cases. If the intent was really to be a warning, than browsers can do that by outputting a warning to the developer console (e.g.`console.warn`). This is already common for other things like unusupported `<meta name="viewport" ...>` values.

---

I'll also reserve discussion about querying for supported types in https://github.com/w3c/performance-timeline/issues/77.

**Comment by igrigorik** [comment]
Minutes from WG call today: https://www.w3.org/2017/09/25-webperf-minutes.html

tl;dr: drop the exception.

- Agreement on the call that raising exception in this case is bad behavior and we should remove it. We may have no-op observers as a result, but that's OK. 
  - We'll add a non-normative note suggesting that UA's can consider adding a console warning when this case is hit. 
- We discussed @wisniewskit's proposal of returning an array from observe() but decided to stay away from it: if you want to detect which types of supported, you have to (a) create the observer and (b) know which properties to enumerate.. which is a lot of work, and we can provide a more ergonomic solution.
  - Consensus on the call is that we should provide a static method on PerformanceObserver that returns the list of supported types -- easy and simple, on both ends.
   - @JosephPecoraro wrt https://github.com/w3c/performance-timeline/issues/77#issuecomment-331809692: the feedback was that it requires know which interfaces to look for, and also that we shouldn't assume that presence of the interface indicates PO support -- e.g. ResourceTiming is supported by some browsers, but not in PO.. yet.

@plehegar has the spec and WPT updates WIP @ https://github.com/w3c/performance-timeline/pull/88. 

If there are any objections or other feedback, please chime in here and/or on the PR.

**Comment by igrigorik** [comment]
Resolved via https://github.com/w3c/performance-timeline/pull/88.
